File: src/line.cc
#include "line.hh"

Line::Line(Chip *chip, unsigned int pin) {
  line = gpiod_chip_get_line(chip->getNativeChip(), pin);
  std::string msg = "Line::new - Unable to open GPIO line ";
  msg = msg + std::to_string(pin);
  if (!line) Nan::ThrowError(Nan::ErrnoException(errno, msg.c_str()));
}

Line::Line(Chip *chip, const char *name) {
  line = gpiod_chip_find_line(chip->getNativeChip(), name);
  std::string msg = "Line::new - Unable to open GPIO line ";
  msg += name;
  if (!line) Nan::ThrowError(Nan::ErrnoException(errno, msg.c_str()));
}

Line::~Line() {
  gpiod_line_release(line);
  line = NULL;
}

gpiod_line *Line::getNativeLine() {
  return line;
}

Nan::Persistent<v8::Function> Line::constructor;

NAN_MODULE_INIT(Line::Init) {
  v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
  tpl->SetClassName(Nan::New("Line").ToLocalChecked());
  tpl->InstanceTemplate()->SetInternalFieldCount(1);
  constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
  Nan::Set(target, Nan::New("Line").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
}

NAN_METHOD(Line::New) {
  if (info.IsConstructCall()) {
    Chip *chip = Nan::ObjectWrap::Unwrap<Chip>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
    Line *obj;
    if (info[1]->IsNumber()) {
      unsigned int pin = Nan::To<unsigned int>(info[1]).FromJust();
      obj = new Line(chip, pin);
    } else {
      Nan::Utf8String name(info[1]);
      obj = new Line(chip, *name);
    }
    obj->Wrap(info.This());
    info.GetReturnValue().Set(info.This());
  } else {
    const int argc = 2;
    v8::Local<v8::Value> argv[argc] = {info[0], info[1]};
    v8::Local<v8::Function> cons = Nan::New(constructor);
    info.GetReturnValue().Set(Nan::NewInstance(cons, argc, argv).ToLocalChecked());
  }
}

NAN_METHOD(getLineOffset) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  int ret = gpiod_line_offset(obj->getNativeLine());
  if (-1 == ret) {
    Nan::ThrowError(Nan::ErrnoException(errno, "::getLineOffset"));
  } else {
    info.GetReturnValue().Set(ret);
  }
}

NAN_METHOD(getLineName) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  const char *name = gpiod_line_name(obj->getNativeLine());
  if (!name) {
    info.GetReturnValue().Set(Nan::Undefined());
  } else {
    info.GetReturnValue().Set(Nan::New<v8::String>(name).ToLocalChecked());
  }
}

NAN_METHOD(getLineConsumer) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  const char *name = gpiod_line_consumer(obj->getNativeLine());
  if (!name) {
    info.GetReturnValue().Set(Nan::Undefined());
  } else {
    info.GetReturnValue().Set(Nan::New<v8::String>(name).ToLocalChecked());
  }
}

NAN_METHOD(getLineDirection) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  int direction = gpiod_line_direction(obj->getNativeLine());
  info.GetReturnValue().Set(direction);
}

NAN_METHOD(getLineActiveState) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  int activeState = gpiod_line_active_state(obj->getNativeLine());
  info.GetReturnValue().Set(activeState);
}

NAN_METHOD(getLineBias) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  int bias = gpiod_line_bias(obj->getNativeLine());
  info.GetReturnValue().Set(bias);
}

NAN_METHOD(isLineUsed) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  bool used = gpiod_line_is_used(obj->getNativeLine());
  info.GetReturnValue().Set(used);
}

NAN_METHOD(isLineFree) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  bool free = gpiod_line_is_free(obj->getNativeLine());
  info.GetReturnValue().Set(free);
}

NAN_METHOD(isLineOpenDrain) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  bool drain = gpiod_line_is_open_drain(obj->getNativeLine());
  info.GetReturnValue().Set(drain);
}

NAN_METHOD(isLineOpenSource) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  bool source = gpiod_line_is_open_source(obj->getNativeLine());
  info.GetReturnValue().Set(source);
}

NAN_METHOD(update) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  if (gpiod_line_update(obj->getNativeLine()) < 0) {
    Nan::ThrowError(Nan::ErrnoException(errno, "line update failed"));
  }
}

NAN_METHOD(needsUpdate) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  bool need = gpiod_line_needs_update(obj->getNativeLine());
  info.GetReturnValue().Set(need);
}

NAN_METHOD(getValue) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  int ret = gpiod_line_get_value(obj->getNativeLine());
  if (-1 == ret) {
    Nan::ThrowError(Nan::ErrnoException(errno, "::getValue"));
  } else {
    info.GetReturnValue().Set(ret);
  }
}

NAN_METHOD(setValue) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  uint32_t value = Nan::To<uint32_t>(info[1]).FromJust();
  if (gpiod_line_set_value(obj->getNativeLine(), value) == -1) {
    Nan::ThrowError(Nan::ErrnoException(errno, "::setValue"));
  }
}

NAN_METHOD(lineRequest) {
  v8::Local<v8::Context> context = Nan::GetCurrentContext();
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(info[0]->ToObject(context).ToLocalChecked());
  struct gpiod_line_request_config config;
  if (!info[1]->IsObject()) {
    Nan::ThrowError(Nan::Error("::lineRequest config is not an object"));
    return;
  }
  v8::Local<v8::Object> jsObj = info[1]->ToObject(context).ToLocalChecked();
  v8::MaybeLocal<v8::Value> consumer = jsObj->Get(context, Nan::New("consumer").ToLocalChecked());
  Nan::Utf8String consumerStr(consumer.ToLocalChecked());
  config.consumer = *consumerStr;
  v8::Local<v8::Value>
      requestType = jsObj->Get(context, Nan::New("requestType").ToLocalChecked()).ToLocalChecked();
  if (requestType.IsEmpty() || !requestType->IsNumber()) {
    Nan::ThrowError(Nan::Error("::lineRequest config.requestType is not a number"));
    return;
  }
  config.request_type = Nan::To<int>(requestType).FromJust();
  config.flags = 0;
  v8::Local<v8::Value> flags = jsObj->Get(context, Nan::New("flags").ToLocalChecked()).ToLocalChecked();
  if (!flags.IsEmpty() && !flags->IsUndefined()) {
    if (!flags->IsNumber()) {
      Nan::ThrowError(Nan::Error("::lineRequest config.flags is not a number"));
      return;
    }
    config.flags = Nan::To<int>(flags).FromJust();
  }
  int defaultValue = 0;
  if (info[2]->IsNumber()) {
    defaultValue = Nan::To<int>(info[2]).FromJust();
  }
  if (gpiod_line_request(obj->getNativeLine(), &config, defaultValue) == -1) {
    Nan::ThrowError(Nan::ErrnoException(errno, "::lineRequest"));
  }
}

NAN_METHOD(requestInputMode) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  if (-1 == gpiod_line_request_input(obj->getNativeLine(), *consumer))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestInputMode"));
}

NAN_METHOD(requestOutputMode) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  unsigned int value = 0;
  v8::Local<v8::Value> defaultValue = info[2];
  if (!defaultValue->IsUndefined() && defaultValue->IsNumber()) {
    unsigned int val = Nan::To<unsigned int>(defaultValue).FromJust();
    if (val > 1) {
      Nan::ThrowError(Nan::ErrnoException(errno, "::requestOutputMode() value is not in {0,1} range"));
      return;
    }
    value = val;
  }
  if (-1 == gpiod_line_request_output(obj->getNativeLine(), *consumer, value))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestOutputMode"));
}

NAN_METHOD(requestRisingEdgeEvents) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  if (-1 == gpiod_line_request_rising_edge_events(obj->getNativeLine(), *consumer))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestRisingEdgeEvents"));
}

NAN_METHOD(requestFallingEdgeEvents) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  if (-1 == gpiod_line_request_falling_edge_events(obj->getNativeLine(), *consumer))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestFallingEdgeEvents"));
}

NAN_METHOD(requestBothEdgesEvents) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  if (-1 == gpiod_line_request_both_edges_events(obj->getNativeLine(), *consumer))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestBothEdgesEvents"));
}

NAN_METHOD(requestInputModeFlags) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  int flags = Nan::To<int>(info[2]).FromJust();
  if (-1 == gpiod_line_request_input_flags(obj->getNativeLine(), *consumer, flags))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestInputModeFlags"));
}

NAN_METHOD(requestOutputModeFlags) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  int flags = Nan::To<int>(info[2]).FromJust();
  int defaultValue = Nan::To<int>(info[3]).FromJust();
  if (-1 == gpiod_line_request_output_flags(obj->getNativeLine(), *consumer, flags, defaultValue))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestOutputModeFlags"));
}

NAN_METHOD(requestRisingEdgeEventFlags) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  int flags = Nan::To<int>(info[2]).FromJust();
  if (-1 == gpiod_line_request_rising_edge_events_flags(obj->getNativeLine(), *consumer, flags))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestRisingEdgeEventFlags"));
}

NAN_METHOD(requestFallingEdgeEventFlags) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  int flags = Nan::To<int>(info[2]).FromJust();
  if (-1 == gpiod_line_request_falling_edge_events_flags(obj->getNativeLine(), *consumer, flags))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestFallingEdgeEventFlags"));
}

NAN_METHOD(requestBothEdgesEventFlags) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  Nan::Utf8String consumer(info[1]);
  int flags = Nan::To<int>(info[2]).FromJust();
  if (-1 == gpiod_line_request_both_edges_events_flags(obj->getNativeLine(), *consumer, flags))
    Nan::ThrowError(Nan::ErrnoException(errno, "::requestBothEdgesEventFlags"));
}

NAN_METHOD(release) {
  Line *obj = Nan::ObjectWrap::Unwrap<Line>(Nan::To<v8::Object>(info[0]).ToLocalChecked());
  gpiod_line_release(obj->getNativeLine());
}

File: lib/line.js
const libgpiod = require('bindings')('node-libgpiod');

/**
 * @typedef {import('./chip').Chip} Chip
 */

/**
 * @class Line
 * @description Class to handle a single line on the chip
 */
class Line {
	/**
	 * Possible values returned by line line direction
	 * @typedef {Object} LineDirection
	 */
	static Direction = {
		INPUT: 1,
		OUTPUT: 2,
	};

	/**
	 * Possible values returned by line active state
	 * @typedef {Object} LineActiveState
	 * @property {number} HIGH - The line is active when high
	 * @property {number} LOW - The line is active when low
	 */
	static ActiveState = {
		HIGH: 1,
		LOW: 2,
	};

	/**
	 * Possible values returned by line bias
	 * @typedef {Object} LineBias
	 * @property {number} AS_IS - The internal bias state is unknown
	 * @property {number} DISABLE - The internal bias is disabled
	 * @property {number} PULL_UP - The internal pull-up bias is enabled
	 * @property {number} PULL_DOWN - The internal pull-down bias is enabled
	 */
	static Bias = {
		AS_IS: 1,
		/** < The internal bias state is unknown. */
		DISABLE: 2,
		/** < The internal bias is disabled. */
		PULL_UP: 3,
		/** < The internal pull-up bias is enabled. */
		PULL_DOWN: 4,
		/** < The internal pull-down bias is enabled. */
	};

	/**
	 * Possible request types for line request
	 */
	static RequestType = {
		get Direction() {
			return {
				/** < Request the line(s), but don't change current direction. */
				AS_IS: 1,
				/** < Request the line(s) for reading the GPIO line state. */
				INPUT: 2,
				/** < Request the line(s) for setting the GPIO line state. */
				OUTPUT: 3,
			};
		},
		get Event() {
			return {
				FALLING_EDGE: 4,
				/** < Only watch falling edge events. */
				RISING_EDGE: 5,
				/** < Only watch rising edge events. */
				BOTH_EDGES: 6,
				/** < Monitor both types of events. */
			};
		},
	};

	/**
	 * Possible flags for line request
	 * @typedef {Object} RequestFlags
	 * @property {number} OPEN_DRAIN - The line is an open-drain port
	 * @property {number} OPEN_SOURCE - The line is an open-source port
	 * @property {number} ACTIVE_LOW - The active state of the line is low (high is the default)
	 * @property {number} BIAS_DISABLE - The line has neither pull-up nor pull-down resistor
	 * @property {number} BIAS_PULL_DOWN - The line has pull-down resistor enabled
	 * @property {number} BIAS_PULL_UP - The line has pull-up resistor enabled
	 */
	static RequestFlags = {
		OPEN_DRAIN: 1,
		/** < The line is an open-drain port. */
		OPEN_SOURCE: 1 << 1,
		/** < The line is an open-source port. */
		ACTIVE_LOW: 1 << 2,
		/** < The active state of the line is low (high is the default). */
		BIAS_DISABLE: 1 << 3,
		/** < The line has neither either pull-up nor pull-down resistor. */
		BIAS_PULL_DOWN: 1 << 4,
		/** < The line has pull-down resistor enabled. */
		BIAS_PULL_UP: 1 << 5,
		/** < The line has pull-up resistor enabled. */
	};

	/**
	 * @constructor
	 * @param {Chip} chip - chip to which the line belongs
	 * @param {string|number} identifier - line number or name
	 */
	constructor(chip, identifier) {
		this._handler = new libgpiod.Line(chip._handler, identifier);
	}

	/**
	 * @property offset - The line number
	 * @returns {number} - The line number
	 *
	 */
	get offset() {
		return libgpiod.getLineOffset(this._handler);
	}

	/**
	 * @property name - The line name
	 * @returns {string} - The line name
	 */
	get name() {
		return libgpiod.getLineName(this._handler);
	}

	/**
	 * @property consumer - Name of line consumer or hog
	 *
	 * @returns {string} - The line hog name, if any
	 */
	get consumer() {
		return libgpiod.getLineConsumer(this._handler);
	}

	/**
	 * @property direction - The line direction
	 * @returns {Line.Direction} - either input (1) or output (2)
	 */
	get direction() {
		return libgpiod.getLineDirection(this._handler);
	}

	/**
	 * @property activeState - The line active state
	 * @returns {Line.ActiveState} - either high (1) or low (2)
	 */
	get activeState() {
		return libgpiod.getLineActiveState(this._handler);
	}

	/**
	 * @property bias - The line bias
	 * @returns {Line.Bias} - either as_is (1), disable (2), pull_up (3) or pull_down (4)
	 */
	get bias() {
		return libgpiod.getLineBias(this._handler);
	}

	/**
	 * @property used - The line used state
	 * @returns {boolean} - true if the line is used, false otherwise
	 */
	get used() {
		return libgpiod.isLineUsed(this._handler);
	}

	/**
	 * @property free - The line free state
	 * @returns {boolean} - true if the line is free, false otherwise
	 */
	get free() {
		return libgpiod.isLineFree(this._handler);
	}

	/**
	 * @property openDrain - The line open drain state
	 * @returns {boolean} - true if the line is open drain, false otherwise
	 */
	get openDrain() {
		return libgpiod.isLineOpenDrain(this._handler);
	}

	/**
	 * @property openSource - The line open source state
	 * @returns {boolean} - true if the line is open source, false otherwise
	 */
	get openSource() {
		return libgpiod.isLineOpenSource(this._handler);
	}

	/**
	 * Update - refresh line state
	 */
	update() {
		return libgpiod.update(this._handler);
	}

	/**
	 * @property needsUpdate - check if line state needs update
	 * @returns {boolean} - true if the line state needs update, false otherwise
	 * @deprecated - does nothing in libgpiod 1.6 series
	 */
	get needsUpdate() {
		return libgpiod.needsUpdate(this._handler);
	}

	/**
	 * GetValue - The line value
	 * @returns {0|1} - The line value, either 0 or 1
	 */
	getValue() {
		return libgpiod.getValue(this._handler);
	}

	/**
	 * SetValue - The line value
	 * @param {number} val - The line value, either 0 or 1
	 */
	setValue(value) {
		return libgpiod.setValue(this._handler, value);
	}

	/**
	 * LineRequest - performs a line request
	 * @param {object} config - request configuration
	 * @param {string} config.consumer - consumer name
	 * @param {Line.RequestType.Direction | Line.RequestType.Event} config.requestType - request type
	 * @param {Line.RequestFlags} config.flags - request flags bits
	 * @param {0|1} [defaultValue] - default value to set
	 */
	lineRequest(config, defaultValue = 0) {
		return libgpiod.lineRequest(this._handler, config, defaultValue);
	}

	/**
	 * RequestInputMode - request line in input mode
	 * @param {string} [consumer] - consumer name
	 */
	requestInputMode(consumer = '') {
		return libgpiod.requestInputMode(this._handler, consumer);
	}

	/**
	 * RequestOutputMode - request line in output mode
	 * @param {string} [consumer] - consumer name
	 * @param {0|1} [defaultValue] - default value to set
	 */
	requestOutputMode(consumer = '', defaultValue = 0) {
		return libgpiod.requestOutputMode(this._handler, consumer, defaultValue);
	}

	/**
	 * RequestRisingEdgeEvents
	 * @param {string} [consumer] - consumer name
	 */
	requestRisingEdgeEvents(consumer = '') {
		return libgpiod.requestRisingEdgeEvents(this._handler, consumer);
	}

	/**
	 * RequestRisingEdgeEvents
	 * @param {string} [consumer] - consumer name
	 */
	requestFallingEdgeEvents(consumer = '') {
		return libgpiod.requestFallingEdgeEvents(this._handler, consumer);
	}

	/**
	 * RequestBothEdgesEvents
	 * @param {string} [consumer] - consumer name
	 */
	requestBothEdgesEvents(consumer = '') {
		return libgpiod.requestBothEdgesEvents(this._handler, consumer);
	}

	/**
	 * RequestInputModeFlags
	 *
	 * @param {string} consumer  - consumer name
	 * @param {Line.RequestFlags} flags  - request flags bits
	 */
	requestInputModeFlags(consumer, flags) {
		return libgpiod.requestInputModeFlags(this._handler, consumer, flags);
	}

	/**
	 * RequestOutputModeFlags
	 * @param {string} consumer  - consumer name
	 * @param {Line.RequestFlags} flags - request flags bits
	 * @param {number} [defaultValue=0] - default value to set
	 */
	requestOutputModeFlags(consumer, flags, defaultValue = 0) {
		return libgpiod.requestOutputModeFlags(this._handler, consumer, flags, defaultValue);
	}

	/**
	 * RequestRisingEdgeEventFlags
	 * @param {string} consumer  - consumer name
	 * @param {Line.RequestFlags} flags  - request flags bits
	 */
	requestRisingEdgeEventFlags(consumer, flags) {
		return libgpiod.requestRisingEdgeEventFlags(this._handler, consumer, flags);
	}

	/**
	 * RequestFallingEdgeEventFlags
	 * @param {string} consumer  - consumer name
	 * @param {Line.RequestFlags} flags  - request flags bits
	 */
	requestFallingEdgeEventFlags(consumer, flags) {
		return libgpiod.requestFallingEdgeEventFlags(this._handler, consumer, flags);
	}

	/**
	 * RequestBothEdgesEventFlags
	 * @param {string} consumer  - consumer name
	 * @param {Line.RequestFlags} flags  - request flags bits
	 */
	requestBothEdgesEventFlags(consumer, flags) {
		return libgpiod.requestBothEdgesEventFlags(this._handler, consumer, flags);
	}

	/**
	 * Release - release the line
	 */
	release() {
		return libgpiod.release(this._handler);
	}
}

exports.Line = Line;

/**
 * Pin sugar for Line. grab lines from chip 0
 * @param {number} number
 */
exports.Pin = function Pin(number) {
	const {Chip} = require('./chip');
	return new Line(new Chip(0), number);
};
